File: Task3/c++/main.cpp
#include <iostream>
#include <sstream>
using namespace std;

// сумма цифр
int sumChis(int n) {
    int sum = 0;
    while (n > 0) {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

// произведение цифр
int ymnChis(int n) {
    int product = 1;
    while (n > 0) {
        product *= n % 10;
        n /= 10;
    }
    return product;
}

int main() {
    int N;
    cout << "Введите количество чисел: ";
    cin >> N;

    if (N <= 0) {
        cout << "Ошибка: N должно быть > 0\n";
        return 1;
    }

    cin.ignore(); // очистка \n после ввода N
    cout << "Введите ваши числа: ";

    string line;
    getline(cin, line); // читаем всю строку

    stringstream ss(line);
    int x, count = 0;
    bool found = false;
    while (ss >> x) {
        count++;

        int sum = sumChis(x);
        int product = ymnChis(x);

        if (sum < product) {
            cout << x << " ";
            found = true;
        }
    }

    // проверка количества
    if (count < N) {
        cout << "\nОшибка: введено меньше чисел, чем " << N << "\n";
        return 1;
    }
    if (count > N) {
        cout << "\nОшибка: введено больше чисел, чем " << N << "\n";
        return 1;
    }

    if (!found) {
        cout << "не найдено";
    }

    return 0;
}

File: Task3/rust/main/src/main.rs
use std::io;

fn sum_digits(mut n: i32) -> i32 {
    let mut sum = 0;
    while n > 0 {
        sum += n % 10;
        n /= 10;
    }
    sum
}

fn product_digits(mut n: i32) -> i32 {
    let mut product = 1;
    while n > 0 {
        product *= n % 10;
        n /= 10;
    }
    product
}

fn main() {
    let mut input = String::new();

    println!("Введите количество чисел:");
    io::stdin().read_line(&mut input).unwrap();
    let n: i32 = input.trim().parse().unwrap();
    println!("Введите ваши числа:");

    let mut found = false;
    let mut count = 0;

    while count < n {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();

        println!("Результат: ");
        for num in input.split_whitespace() {
            if count >= n {
                break;
            }

            let x: i32 = num.parse().unwrap();
            let sum = sum_digits(x);
            let product = product_digits(x);

            if sum < product {
                print!("{} ", x);
                found = true;
            }

            count += 1;
        }
    }

    if !found {
        println!("не найдено");
    }
}
